### PR TITLE
CMR-4145: Added variable association endpoint in search.

### PR DESCRIPTION
**Pull request by ygliuvt**
This is part 4 of a four-parter for creating variable associations with collections:

1. Add creating variable association concept in metadata-db
2. Add search of variables in metadata-db
3. Add search of variable associations in metadata-db
4. Add variable associations endpoint in search-app

The integration tests for this issue will be added once we have the endpoint of ingest of variable concept in place in ingest-app.

### CODE REVIEW

File: search-app/src/cmr/search/api/variables_api.clj [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    (context "/:variable-name" [variable-name]
+      (context "/associations" []
+
+        ;; Associate a variable with a list of collections
+        (POST "/" {:keys [request-context headers body]}
+          (associate-variable-to-collections
+           request-context headers (slurp body) (str/lower-case variable-name)))
+
+        ;; Dissociate a variable from a list of collections
```
Comment (chris-durbin) [inline search-app/src/cmr/search/api/variables_api.clj]: We have a separate ticket to address this - is this implemented as well?

File: search-app/src/cmr/search/services/association_validation.clj [hunk 2]
```diff
@@ -0,0 +1,260 @@
+(ns cmr.search.services.association-validation
+  "This contains functions for validating the business rules of an association between a collection
+   and a tag or a collection between a variable."
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/association_validation.clj]: Reword "collection between a variable"

File: search-app/src/cmr/search/services/association_validation.clj [hunk 3]
```diff
@@ -0,0 +1,260 @@
+(ns cmr.search.services.association-validation
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/association_validation.clj]: I'm glad you were able to reuse a large chunk of code for both types of associations.

File: search-app/src/cmr/search/services/association_validation.clj [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+                            :result-format :query-specified
+                            :result-fields [:concept-id]
+                            :skip-acls? false})
+          concept-ids (->> (qe/execute-query context query)
+                           :items
+                           (map :concept-id))]
+      (set/difference (set coll-concept-ids) (set concept-ids)))))
+
+(defn- get-bad-collection-revisions
+  "Returns the bad collection revisions of the given tag associations partitioned into a set of
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/association_validation.clj]: Remove the word "tag" from all of the docstrings that apply to both tag and variable associations. There are a bunch.

File: search-app/src/cmr/search/services/association_validation.clj [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+      :else
+      (format (str "Tag can only be associated with a collection or a collection revision, "
+                   "never both at the same time. There are already conflicting tag associations "
+                   "in metadata-db with tag key [%s] on collection [%s] , please delete "
+                   "one of the conflicting tag associations.")
+              tag-key (:concept-id tag-association)))))
+
+(defmethod validate-association-conflict-for-collection :variable
+  [context assoc-type variable-name variable-association]
+  (let [tas (->> (mdb/find-concepts context
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/association_validation.clj]: There's a lot of common code - not sure if you want to try and refactor the common into functions that both of these multimethods used. Also bindings like `tas` should reflect variable associations.

File: search-app/src/cmr/search/services/variable_service.clj [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+(defn- context->user-id
+  "Returns user id of the token in the context. Throws an error if no token is provided"
+  [context]
+  (if-let [token (:token context)]
+    (util/lazy-get context :user-id)
+    (errors/throw-service-error
+     :unauthorized "Variables cannot be modified without a valid user token.")))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Public API
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/variable_service.clj]: The public API actually starts on line 203.

File: system-int-test/test/cmr/system_int_test/search/tagging/tag_association_collection_revisions_test.clj [hunk 7]
```diff
@@ -56,107 +56,108 @@
 
     (testing "successful case, the tag association keys can have either _ or -"
       (let [response (tags/associate-by-concept-ids
-                       token tag-key [{:concept_id (:concept-id coll1-1)
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/tagging/tag_association_collection_revisions_test.clj]: James is going to be very unhappy about all of these lines modified to remove a space.

File: search-app/src/cmr/search/api/variables_api.clj [hunk 8]
```diff
@@ -0,0 +1,92 @@
+(ns cmr.search.api.variables-api
+  "Defines the API for variableging collections in the CMR."
```
Comment (ygliuvt) [inline search-app/src/cmr/search/api/variables_api.clj]: Fixed.

File: search-app/src/cmr/search/api/variables_api.clj (hunk 1 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/api/variables_api.clj]: Yes.

File: search-app/src/cmr/search/services/association_validation.clj (hunk 2 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/association_validation.clj]: Done.

File: search-app/src/cmr/search/services/association_validation.clj (hunk 4 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/association_validation.clj]: Done.

File: search-app/src/cmr/search/services/association_validation.clj (hunk 5 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/association_validation.clj]: I have tried to refactor duplicated code into common functions, and these functions are left alone as they are more readable this way. I will rename the `tas`s.

File: search-app/src/cmr/search/services/variable_service.clj (hunk 6 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/variable_service.clj]: Removed unnecessary comment.

File: system-int-test/test/cmr/system_int_test/search/tagging/tag_association_collection_revisions_test.clj (hunk 7 above)
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/tagging/tag_association_collection_revisions_test.clj]: Life is so hard!

**Pull request merged**